src/precomputed_quantities.jl

```diff
@@ -111,16 +134,218 @@ function ts(Yc, K, Φ, thermo_params, energy_form, moisture_model)
+            (; q_pt = TD.PhasePartition(q_totʲ, q_liqʲ, q_iceʲ))
+        end
+    end
+    return thermo_state(thermo_params; p, energy_var..., moisture_var...)
+end
+
+
+"""
+    set_precomputed_quantities!(Y, p, t)
+
+Uses the prognostic state `Y` to set the values of `ᶜu`, `ᶠu³`, `ᶜK`, `ᶜts`, and
+`ᶜp`. If `turbconv_model` is `EDMFX`, also sets the values of `ᶜuʲs`, `ᶠu³ʲs`,
+`ᶜKʲs`, `ᶜtsʲs`, `ᶜρa⁰`, `ᶠw⁰`, `ᶜu⁰`, `ᶠu³⁰`, `ᶜK⁰`, and `ᶜts⁰`.
+
+This function also applies a "filter" to `Y` in order to ensure that `ᶠu³` is 0
+at the surface (i.e., to enforce the impenetrable boundary condition). If
+`turbconv_model` is `EDMFX`, the filter also ensures that the `ᶠu³ʲs` and `ᶠu³⁰`
+are 0 at the surface. In the future, we may want to move this filtering
+elsewhere, but doing it here ensures that it occurs whenever the precomputed
+quantities are updated.
+
+Note: If you need to update the value of any of the aforementioned quantities,
+please call this function instead of recomputing the value yourself. Otherwise,
+it will be difficult to ensure that the duplicated computations are consistent.
+"""
 function set_precomputed_quantities!(Y, p, t)
-    (; energy_form, moisture_model) = p.atmos
+    (; energy_form, moisture_model, turbconv_model) = p.atmos
     thermo_params = CAP.thermodynamics_params(p.params)
     (; ᶜΦ) = p
+    ᶠinterp = Operators.InterpolateC2F(
+        bottom = Operators.Extrapolate(),
+        top = Operators.Extrapolate(),
+    )
 
     # TODO: This should probably be moved to dss! (i.e., enforce_constraints!).
-    set_boundary_velocity!(Y)
+    set_boundary_velocity!(Y, turbconv_model)
 
     (; ᶜu, ᶠu³, ᶜK, ᶜts, ᶜp) = p
     set_velocity_quantities!(ᶜu, ᶠu³, ᶜK, Y.f.w, Y.c)
+    if turbconv_model isa EDMFX
+        @. ᶜK += Y.c.sgs⁰.ρatke / Y.c.ρ
+        # TODO: We actually need to add ᶜρatke⁰ / ᶜρ⁰ to ᶜK, not ᶜρatke⁰ / ᶜρ.
+        # However, we cannot compute ᶜρ⁰ without first computing ᶜts⁰, which
+        # in turn depends on ᶜp, which depends on ᶜts, which depends on ᶜK. So,
+        # unless we decide to run a nonlinear solver, this circular dependency
+        # will prevent us from computing the exact value of ᶜK. For now, we will
+        # make the anelastic approximation ᶜρ⁰ ≈ ᶜρ.
+    end
     @. ᶜts = ts(Y.c, ᶜK, ᶜΦ, thermo_params, energy_form, moisture_model)
     @. ᶜp = TD.air_pressure(thermo_params, ᶜts)
+
+    if turbconv_model isa EDMFX
+        (; a_min) = turbconv_model
+        (; ᶜuʲs, ᶠu³ʲs, ᶜKʲs, ᶜtsʲs, ᶜρa⁰, ᶠw⁰, ᶜu⁰, ᶠu³⁰, ᶜK⁰, ᶜts⁰) = p
+
+        tsʲs(Yc, Kʲs, p, Φ) =
+            map(Yc.sgsʲs, Kʲs) do sgsʲ, Kʲ
+                tsʲ(
+                    sgsʲ,
+                    Yc,
+                    Kʲ,
+                    p,
+                    Φ,
+                    a_min,
+                    thermo_params,
+                    energy_form,
+                    moisture_model,
+                )
+            end
+
+        w⁰(ρa⁰, ρaʲs, wʲs, ρ, w) =
```

Comment:
The \^0 looks more like a B on my screen. Not sure how does it render for others?